serverless/internal/log/integrate.go

```diff
@@ -17,25 +17,150 @@
 package log
 
 import (
-	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
 
+	"github.com/golang/glog"
 	"github.com/google/trillian-examples/serverless/api"
+	"github.com/google/trillian-examples/serverless/internal/storage"
+	"github.com/google/trillian/merkle/compact"
 	"github.com/google/trillian/merkle/hashers"
 )
 
 // Storage represents the set of functions needed by the log tooling.
 type Storage interface {
+	// GetTile returns the tile at the given level & index.
+	GetTile(level, index, logSize uint64) (*api.Tile, error)
+
+	// StoreTile stores the tile at the given level & index.
+	StoreTile(level, index uint64, tile *api.Tile) error
+
 	// LogState returns the current state of the stored log.
 	LogState() api.LogState
 
 	// UpdateState stores a newly updated log state.
 	UpdateState(newState api.LogState) error
 
+	// ScanSequenced calls f for each contiguous sequenced log entry >= begin.
+	// It should stop scanning if the call to f returns an error.
+	ScanSequenced(begin uint64, f func(seq uint64, entry []byte) error) (uint64, error)
+
 	// Sequence assigns sequence numbers to the passed in entry.
 	Sequence(leafhash []byte, leaf []byte) error
 }
 
 // Integrate adds sequenced but not-yet-included entries into the tree state.
 func Integrate(st Storage, h hashers.LogHasher) error {
-	return errors.New("unimplemented")
+	rf := compact.RangeFactory{Hash: h.HashChildren}
+
+	// fetch state
+	state := st.LogState()
+	baseRange, err := rf.NewRange(0, state.Size, state.Hashes)
+	if err != nil {
+		return fmt.Errorf("failed to create range covering existing log: %w", err)
+	}
+
+	r, err := baseRange.GetRootHash(nil)
+	if err != nil {
+		return fmt.Errorf("invalid log state, unable to recalculate root: %w", err)
+	}
+	glog.Infof("Loaded state with roothash %x", r)
+
+	tiles := make(map[string]*api.Tile)
+
+	visitor := func(id compact.NodeID, hash []byte) {
```

Comment:
Some documentation on this would be useful. There's a lot going on here as this is binding a) the visitor pattern from compact range, b) the in-memory `tiles` map, c) the actual storage.

The upshot of what it's trying to do from what I can see is: populate `tiles` with all the tiles that need to be updated. This creates new tiles if needed, and updates any existing tiles already in the storage. The tiles will not be written back to storage, but will be stored in `tiles` afterwards.

Reply:
I've factored out the cache/visitor and added a bunch of comments throughout.